app/common/console_output.py

```diff
@@ -68,34 +85,33 @@ def _parse_from_offset(self, max_lines, offset_line):
 
         return ConsoleOutputSegment(offset_line, output_lines, total_lines, ''.join(console_output))
 
-    def _parse_from_end(self, max_lines):
+    def _parse_from_end(self, max_lines: int) -> ConsoleOutputSegment:
         """
         Return console output segment containing the last max_lines of output, or the whole file
         if less than or equal to max_lines.
-
-        :type max_lines: int
-        :rtype: ConsoleOutputSegment
         """
-        total_lines = 0
-        output_lines = 0
-        console_output = []
+        total_num_lines = 0
+        offset_line = 0
+        console_output_lines = collections.deque()
 
-        with open(self.path, 'r', encoding='utf-8', errors='replace') as f:
-            # Figure out how many lines there are, so that we know which offset line to start appending
-            # console output from.
+        with self._file as f:
             for line in f:
-                if line.endswith("\n"):
-                    total_lines += 1
+                if not line.endswith(b'\n'):
+                    break  # last line; file still (probably) being written to
 
-            f.seek(0)
```

Comment:
Zipfile does not support `seek` so this method had to be rewritten to use a different approach.